File: Dec_1/calorie_count.py
"""
--- Day 1: Calorie Counting ---
Find the Elf carrying the most Calories. How many total Calories is that Elf carrying?

"""
# solution below is )(n) in time complexity and space complexity
# if I had not tracked the elf ids it could have had better space compleity

# opening the file in a way that automatically closes it
with open("/Users/fleigh/Projects/AdventofCode/DecOne/Data.txt", "r") as f:
  # reading the data from the file
  file_data = f.read()
  # splitting the file data into a list with inclusion of empty lines
  cals_list_by_elf = file_data.splitlines()

# we will keep track of both the cals and which elves have the cals


def maxCals(cals_list):
    # initialize elf counter and cache
    curr_elf_id = 1
    elf_dict = {}
    max_cals = 0

    # iterate through the list
    for i in cals_list:
        # when empty space, increment current elf id
        if i == '':
            curr_elf_id += 1
        # otherwise, add to that elf's calorie sum in the elf dict (gotta convert the type to int)
        else:
            elf_dict[curr_elf_id] = elf_dict.get(curr_elf_id, 0) + int(i)
            # and check if this elf has the max cals so far
            max_cals = max(max_cals, elf_dict[curr_elf_id])

    # identify which elf/elves has the value matching our max cals count
    elf_with_max_cals = [k for k, m in elf_dict.items() if m == max_cals]

    # return the elf id and calorie amount
    return (elf_with_max_cals, max_cals)


# call function and print to terminal
print(maxCals(cals_list_by_elf))


"""
--- Part Two ---
Find the top three Elves carrying the most Calories. How many Calories are those Elves carrying in total?

"""
# this solution in O(nlogn) in time compleity due to the addition of sorting
# still order n space complexity

# opening the file in a way that automatically closes it
with open("/Users/fleigh/Projects/AdventofCode/DecOne/Data.txt", "r") as f:
  # reading the data from the file
  file_data = f.read()
  # splitting the file data into a list with inclusion of empty lines
  cals_list_by_elf = file_data.splitlines()

# we will keep track of both the cals AND which elves have the cals

def maxCals2(cals_list):
    # initialize elf counter and cache
    curr_elf_id = 1
    elf_dict = {}

    # iterate through the list
    for i in cals_list:
        # when empty space, increment current elf id
        if i == '':
            curr_elf_id += 1
        # otherwise, add to that elf's calorie sum (gotta convert the type to int)
        else:
            elf_dict[curr_elf_id] = elf_dict.get(curr_elf_id, 0) + int(i)

    # convert dict into list of tuples sorted by the calorie values
    sorted_elves_by_cals = sorted(
        elf_dict.items(), key=lambda x: x[1], reverse=True)

    # return array of 3 tuples containing the 3 elf ids (i=0) amd the largest calorie amounts (i=1)
    top_three_elves = sorted_elves_by_cals[0:3]

    # calculate sum of cals for the 3 elves
    sum_cals_top_three = 0
    for elf in top_three_elves:
        sum_cals_top_three += elf[1]

    # return the top 3 elves and the sum of their cals
    return top_three_elves, sum_cals_top_three


# call function and print to terminal
print(maxCals2(cals_list_by_elf))

File: Dec_1/calorieCount.js
/*
--- Day 1: Calorie Counting ---
Find the Elf carrying the most Calories. How many total Calories is that Elf carrying?
*/

// the question says to keep track of which elves has the most calories
// import Node's file reader
const fs = require("fs");
// read contents of the file
const data = fs.readFileSync("./Data.txt", "UTF-8");

// split the contents by new line into array of substrings
// A carriage return (\r) + newline (\n) is best practice
const array = data.split(/\r?\n/);

const calCount = (arrayOfData) => {
  // to parse string into elves and their calorie counts
  // initialize the elf cache, elf count, and max calorie count
  const elves = {};
  let currElf = 1;
  let maxCals = 0;

  arrayOfData.forEach((element) => {
    // if the element is an eplty string, switch to the next elf
    if (!element) currElf++;
    if (elves[currElf]) {
      elves[currElf] = elves[currElf] + Number(element);
    } else {
      elves[currElf] = Number(element);
    }
    maxCals = Math.max(maxCals, elves[currElf]);
  });

  // use filter method in case more than one elf ties for the max calories brought
  const elfWithMaxCals = Object.keys(elves).filter((key) => elves[key] === maxCals);

  return [
    "Elf(s) with most calories: ",
    elfWithMaxCals,
    "max calorie amount: ",
    maxCals,
  ];
};

console.log(calCount(array));

/*
--- Part Two ---
Find the top three Elves carrying the most Calories. How many Calories are those Elves carrying in total?
*/

// here we will also keep track of which elves as well as how many cals
const calCount2 = (arrayOfData) => {
  // to parse string into elves and their calorie counts
  // initialize the elf cache, elf count, and max calorie count
  const elves = {};
  let currElf = 1;

  arrayOfData.forEach((element) => {
    // if the element is an eplty string, switch to the next elf
    if (!element) currElf++;
    if (elves[currElf]) {
      elves[currElf] = elves[currElf] + Number(element);
    } else {
      elves[currElf] = Number(element);
    }
  });

  // sort the calories values in accending order
  const calsArray = Object.values(elves).sort((a, b) => a - b);
  // isolate the top 3 values
  const topThree = calsArray.slice(-3);
  // use reduce to add the top 3 values together
  const maxCals = topThree.reduce((a, b) => a + b);

  // use filter method to identify the top 3+ elves by their calorie amounts
  const maxElves = Object.keys(elves).filter(
    (key) =>
      elves[key] === topThree[0] ||
      elves[key] === topThree[1] ||
      elves[key] === topThree[2]
  );

  return [
    "Elves with the most calories: ",
    maxElves,
    "max calorie amount: ",
    maxCals,
  ];
};

console.log(calCount2(array));
